src/dsl.ts

```diff
@@ -637,86 +655,152 @@ const permission = seqMap(
     };
   },
 );
 
+// Import expressions
+const importExpression: Parser<ParsedImportExpression> = seqMap(
+  index,
+  seq(lexeme(string("import")), stringLiteral, terminator.atMost(1)),
+  index,
+  function (startIndex, [_, path, __], endIndex) {
+    return {
+      kind: "import",
+      path,
+      range: { startIndex, endIndex },
+    };
+  },
+);
+
+function isRelation(member: DefinitionMember): member is ParsedRelation {
+  return member.kind === "relation";
+}
+
+function isPermission(member: DefinitionMember): member is ParsedPermission {
+  return member.kind === "permission";
+}
+
+function isPartialReference(
+  member: DefinitionMember,
+): member is PartialReference {
+  return member.kind === "partialreference";
+}
+
 // Object Definitions.
-const definition = seqMap(
+const definition: Parser<ParsedObjectDefinition> = seqMap(
   index,
   seq(
     lexeme(string("definition")),
     path,
-    lbrace.then(relationOrPermission.atLeast(0)).skip(rbrace),
+    lbrace.then(definitionMember.atLeast(0)).skip(rbrace),
   ),
   index,
   function (startIndex, data, endIndex) {
-    const rp = data[2] as Array<RelationOrPermission>;
+    const members = data[2];
     return {
       kind: "objectDef",
       name: data[1],
-      relations: rp.filter(
-        (relOrPerm: RelationOrPermission) => "allowedTypes" in relOrPerm,
-      ),
-      permissions: rp.filter(
-        (relOrPerm: RelationOrPermission) => !("allowedTypes" in relOrPerm),
-      ),
+      relations: members.filter(isRelation),
+      permissions: members.filter(isPermission),
+      partialReferences: members.filter(isPartialReference),
+      range: { startIndex: startIndex, endIndex: endIndex },
+    };
+  },
+);
+
+// Partial Definitions.
+// NOTE: these are parsed the same as definitions.
+// TODO: see if this can be combined with objectDef in a sane way
+const partial: Parser<ParsedPartialDefinition> = seqMap(
+  index,
+  seq(
+    lexeme(string("partial")),
+    path,
+    lbrace.then(definitionMember.atLeast(0)).skip(rbrace),
+  ),
+  index,
+  function (startIndex, data, endIndex) {
+    const members = data[2];
+    return {
+      kind: "partial",
+      name: data[1],
+      relations: members.filter(isRelation),
+      permissions: members.filter(isPermission),
+      partialReferences: members.filter(isPartialReference),
       range: { startIndex: startIndex, endIndex: endIndex },
     };
   },
 );
 
 // Caveats.
-const caveatParameterTypeExpr: Parser<unknown> = lazy(() => {
-  return seqMap(
-    index,
-    seq(
-      identifier,
-      lcaret.then(caveatParameterTypeExpr).skip(rcaret).atMost(1),
-    ),
-    index,
-    function (startIndex, data, endIndex) {
-      return {
-        kind: "caveatParameterTypeExpr",
-        name: data[0],
-        generics: data[1],
-        range: { startIndex: startIndex, endIndex: endIndex },
-      };
-    },
-  );
-});
+const caveatParameterTypeExpr: Parser<ParsedCaveatParameterTypeRef> = lazy(
```

Comment:
The only thing that changed here is the type on the parser, but prettier reordered things so it looks like more changed.